kafka-streams-framework/src/main/java/org/hypertrace/core/kafkastreams/framework/callbacks/AbstractCallbackRegistryPunctuator.java

```diff
@@ -10,48 +11,54 @@
 import org.hypertrace.core.kafkastreams.framework.callbacks.action.CallbackAction;
 
 public abstract class AbstractCallbackRegistryPunctuator<T> implements Punctuator {
-  private final KeyValueStore<Long, List<T>> objectStore;
+  private final Clock clock;
+  private final KeyValueStore<Long, ArrayList<T>> objectStore;
   private final CallbackRegistryPunctuatorConfig config;
 
   public AbstractCallbackRegistryPunctuator(
-      CallbackRegistryPunctuatorConfig config, KeyValueStore<Long, List<T>> objectStore) {
+      Clock clock,
+      CallbackRegistryPunctuatorConfig config,
+      KeyValueStore<Long, ArrayList<T>> objectStore) {
+    this.clock = clock;
     this.config = config;
     this.objectStore = objectStore;
   }
 
   public void invoke(long atTimestampInMs, T object) {
     long windowAlignedTimestamp = getWindowAlignedTimestamp(atTimestampInMs);
-    List<T> objectsAtWindow =
+    ArrayList<T> objectsAtWindow =
         Optional.ofNullable(objectStore.get(windowAlignedTimestamp)).orElse(new ArrayList<>());
     objectsAtWindow.add(object);
     objectStore.put(windowAlignedTimestamp, objectsAtWindow);
   }
 
   public void cancelInvocation(long atTimestampInMs, T object) {
     long windowAlignedTimestamp = getWindowAlignedTimestamp(atTimestampInMs);
-    List<T> objectsAtWindow =
+    ArrayList<T> objectsAtWindow =
         Optional.ofNullable(objectStore.get(windowAlignedTimestamp)).orElse(new ArrayList<>());
     objectsAtWindow.remove(object);
-    if (!objectsAtWindow.isEmpty()) {
+    if (objectsAtWindow.isEmpty()) {
+      objectStore.delete(windowAlignedTimestamp);
```

Comment:
what's object stores contract on delete if not exists (since it's possible the empty list came from the orElse case).

Reply:
```
Delete the value from the store (if there is one).
Params:
key – The key
Returns:
The old value or null if there is no such key.
Throws:
NullPointerException – If null is used for key.
V delete(K key);
```